internal/charmstore/resources_test.go

```diff
@@ -0,0 +1,256 @@
+
+	checkResourceDocs(c, docs, expected)
+}
+
+func (s *resourceSuite) TestListResourcesCharmWithoutResources(c *gc.C) {
+	store := s.newStore(c, false)
+	defer store.Close()
+	channel := params.StableChannel
+	curl := charm.MustParseURL("cs:~charmers/precise/wordpress-23")
+	entity, _ := addCharm(c, store, curl)
+
+	resources, err := store.ListResources(entity, channel)
+	c.Assert(err, jc.ErrorIsNil)
+
+	c.Check(resources, gc.HasLen, 0)
+}
+
+func (s *resourceSuite) TestListResourcesBundle(c *gc.C) {
+	store := s.newStore(c, false)
+	defer store.Close()
+	channel := params.StableChannel
+	curl := charm.MustParseURL("cs:~charmers/bundle/wordpress-simple-0")
+	entity := addBundle(c, store, curl)
+
+	_, err := store.ListResources(entity, channel)
+
+	c.Check(err, gc.ErrorMatches, `bundles do not have resources`)
+}
+
+func (s *resourceSuite) TestListResourcesResourceNotFound(c *gc.C) {
+	store := s.newStore(c, false)
+	defer store.Close()
+	channel := params.StableChannel
+	curl := charm.MustParseURL("cs:~charmers/xenial/starsay-3")
+	entity, ch := addCharm(c, store, curl)
+	expected := extractResources(c, curl, ch)
+	sortResources(expected)
+	expected[0] = &mongodoc.Resource{
+		BaseURL: expected[0].BaseURL,
+		Name:    expected[0].Name,
+	}
+	expected[1].Revision = addResource(c, store, entity, channel, expected[1], nil)
+	expected[2].Revision = addResource(c, store, entity, channel, expected[2], nil)
+
+	docs, err := store.ListResources(entity, channel)
+	c.Assert(err, jc.ErrorIsNil)
+
+	checkResourceDocs(c, docs, expected)
+}
+
+func addResources(c *gc.C, store *Store, curl *charm.URL, channel params.Channel, entity *mongodoc.Entity, ch *charm.CharmDir) []*mongodoc.Resource {
+	docs := extractResources(c, curl, ch)
+	for i, doc := range docs {
+		meta := ch.Meta().Resources[doc.Name]
+		blob, err := os.Open(filepath.Join(ch.Path, meta.Path))
+		c.Assert(err, jc.ErrorIsNil)
+		docs[i].Revision = addResource(c, store, entity, channel, doc, blob)
+	}
+	return docs
+}
+
+func addResource(c *gc.C, store *Store, entity *mongodoc.Entity, channel params.Channel, doc *mongodoc.Resource, blobReader io.Reader) int {
+	revision := doc.Revision + 1
+	if blobReader != nil {
+		blob := ResourceBlob{
+			Reader: blobReader,
+			Hash:   doc.BlobHash,
+			Size:   doc.Size,
+		}
+		err := store.addResource(entity, doc.Name, blob, revision)
+		c.Assert(err, jc.ErrorIsNil)
+	} else {
+		doc.Revision = revision
+		err := store.insertResource(doc)
+		c.Assert(err, jc.ErrorIsNil)
+	}
+	if channel != params.UnpublishedChannel {
+		err := store.setResource(entity, channel, doc.Name, revision)
+		c.Assert(err, jc.ErrorIsNil)
+	}
+	return revision
+}
+
+func extractResources(c *gc.C, cURL *charm.URL, ch *charm.CharmDir) []*mongodoc.Resource {
+	copied := *cURL
+	cURL = &copied
+	cURL.Revision = -1
+	cURL.Series = ""
+	var docs []*mongodoc.Resource
+	for _, meta := range ch.Meta().Resources {
+		data, err := ioutil.ReadFile(filepath.Join(ch.Path, meta.Path))
```

Comment:
I don't get this. The Path attribute is supposed to represent where
the resource is stored inside the runtime charm data directory, not
the actual charm itself. Storing the resources inside the charm itself
seems kinda wrong. Also, this looks like we're entrenching the pattern
of using charm dirs, where we'd really like to move away from them
towards in-memory charms etc.

I'd prefer to see some in-memory solution, perhaps fabricating known
contents for the resources, or using a function argument to make that
pluggable if needed.